src/DateTimeParser.jl

```diff
@@ -0,0 +1,512 @@
+            end
+        end
+    end
+
+    # Process year/month/day
+    len_ymd = length(ymd)
+
+    if len_ymd > 3
+        # More than three members!?
+        error("Failed to parse date")
+    elseif len_ymd == 1 || (mstridx != -1 && len_ymd == 2)
+        # One member, or two members with a month string
+        if mstridx != -1
+            res["month"] = ymd[mstridx]
+            deleteat!(ymd, mstridx)
+        end
+        if len_ymd > 1 || mstridx == -1
+            if ymd[1] > 31
+                res["year"] = ymd[1]
+            else
+                res["day"] = ymd[1]
+            end
+        end
+    elseif len_ymd == 2
+        # Two members with numbers
+        if ymd[1] > 31
+            # 99-01
+            res["year"], res["month"] = ymd
+        elseif ymd[2] > 31
+            # 01-99
+            res["month"], res["year"] = ymd
+        elseif dayfirst && ymd[2] <= 12
+            # 13-01
+            res["day"], res["month"] = ymd
+        else
+            # 01-13
+            res["month"], res["day"] = ymd
+        end
+    elseif len_ymd == 3
+        # Three members
+        if mstridx == 1
+            res["month"], res["day"], res["year"] = ymd
+        elseif mstridx == 2
+            if ymd[1] > 31 || (yearfirst && ymd[3] <= 31)
+                # 99-Jan-01
+                res["year"], res["month"], res["day"] = ymd
+            else
+                # 01-Jan-01
+                # Give precendence to day-first, since
+                # two-digit years is usually hand-written.
+                res["day"], res["month"], res["year"] = ymd
+            end
+        elseif mstridx == 3
+            # WTF
+            if ymd[2] > 31
+                # 01-99-Jan
+                res["day"], res["year"], res["month"] = ymd
+            else
+                res["year"], res["day"], res["month"] = ymd
+            end
+        else
+            if ymd[1] > 31 || (yearfirst && ymd[2] <= 12 && ymd[3] <= 31)
+                # 99-01-01
+                res["year"], res["month"], res["day"] = ymd
+            elseif ymd[1] > 12 || (dayfirst && ymd[2] <= 12)
+                # 13-01-01
+                res["day"], res["month"], res["year"] = ymd
+            else
+                # 01-13-01
+                res["month"], res["day"], res["year"] = ymd
+            end
+        end
+    end
+
+    # Fill in default values if none exits
+    if !haskey(res, "year") && !haskey(res, "month") && !haskey(res, "day")
+        res["year"], res["month"], res["day"] = Dates.yearmonthday(default)
+    else
+        if haskey(res, "year")
+            res["year"] = convertyear(res["year"])
+        else
+            res["year"] = year(default)
+        end
+        if !haskey(res, "month")
+            res["month"] = 1
+        end
+        if !haskey(res, "day")
+            res["day"] = 1
+        end
+    end
+    if !haskey(res, "hour") && !haskey(res, "minute") && !haskey(res, "second") && !haskey(res, "millisecond")
```

Comment:
Simplify: `!issubset(keys(res), ("hour", "minute", "second", "millisecond"))`. Alternatively you could do: `any(k -> ("hour", "minute", "second", "millisecond"), keys(res))`. Probably other options too